pkg/provisioning/import.go

```diff
@@ -0,0 +1,712 @@
+	"context"
+	"fmt"
+
+	"github.com/conduitio/conduit/pkg/connector"
+	"github.com/conduitio/conduit/pkg/foundation/cerrors"
+	"github.com/conduitio/conduit/pkg/pipeline"
+	"github.com/conduitio/conduit/pkg/processor"
+	"github.com/conduitio/conduit/pkg/provisioning/config"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+)
+
+// Import takes a pipeline config and imports it into Conduit.
+func (s *Service) Import(ctx context.Context, newConfig config.Pipeline) error {
+	oldConfig, err := s.Export(ctx, newConfig.ID)
+	if err != nil && !cerrors.Is(err, pipeline.ErrInstanceNotFound) {
+		s.logger.Warn(ctx).Err(err).Msgf("could not export pipeline with ID %v, trying to provision new pipeline despite the error", newConfig.ID)
+		// TODO does this behavior make sense? Should we rather return the error
+		//  and skip provisioning?
+	}
+
+	var actions []action
+	actions = append(actions, s.cascadingActionsDeleteOld(oldConfig, newConfig)...)
+	actions = append(actions, s.cascadingActionsUpsertNew(oldConfig, newConfig)...)
+
+	lastActionIndex, err := s.executeActions(ctx, actions)
+	if err != nil {
+		rollbackActions := actions[:lastActionIndex+1]
+		s.logger.Debug(ctx).Err(err).Msgf("rolling back %d import actions", len(rollbackActions))
+		s.reverseActions(rollbackActions) // execute rollback actions in reversed order
+		if ok := s.rollbackActions(ctx, rollbackActions); !ok {
+			s.logger.Warn(ctx).Msg("some actions failed to be rolled back, Conduit state might be corrupted; please report this issue to the Conduit team")
+		}
+		return err
+	}
+
+	return nil
+}
+
+// prepareActionsNewConfig prepares actions needed for the new config to be
+// provisioned. It either creates or updates existing entities.
+func (s *Service) cascadingActionsUpsertNew(oldConfig, newConfig config.Pipeline) []action {
+	var actions []action
+
+	actions = append(actions, s.preparePipelineActions(oldConfig, newConfig)...)
+
+	for _, connNewConfig := range newConfig.Connectors {
+		connOldConfig, _ := s.findConnectorByID(oldConfig.Connectors, connNewConfig.ID)
+		actions = append(actions, s.prepareConnectorActions(connOldConfig, connNewConfig, newConfig.ID)...)
+
+		parent := processor.Parent{
+			ID:   connNewConfig.ID,
+			Type: processor.ParentTypeConnector,
+		}
+		for _, procNewConfig := range connNewConfig.Processors {
+			procOldConfig, _ := s.findProcessorByID(connOldConfig.Processors, procNewConfig.ID)
+			actions = append(actions, s.prepareProcessorActions(procOldConfig, procNewConfig, parent)...)
+		}
+	}
+
+	parent := processor.Parent{
+		ID:   newConfig.ID,
+		Type: processor.ParentTypePipeline,
+	}
+	for _, procNewConfig := range newConfig.Processors {
+		procOldConfig, _ := s.findProcessorByID(oldConfig.Processors, procNewConfig.ID)
+		actions = append(actions, s.prepareProcessorActions(procOldConfig, procNewConfig, parent)...)
+	}
+	return actions
+}
+
+// prepareActionsOldConfig prepares actions needed to delete existing entities
```

Comment:
Thanks for noticing, I'll correct the comment! The name sucks though, let me know if you by any chance think of a better one 🙂

Reply:
Besides maybe `prepareDeleteActions` (which isn't really amazing), I can't think of anything else tbh.

Reply:
I'd like to avoid "prepare", the functions below are already using that word (e.g. `preparePipelineActions`) and those only prepare actions for one entity. I wanted to make a distinction in the names for these two functions, as these are actually diving into all entities and call the "prepare actions" functions (that's why I used "cascading", "traverse" was another option I was thinking of).

That said, I'm ok with using "prepare" if we distinguish these from the other functions 🙂 Also I'm fine with renaming the other functions if needed.

Reply:
I refactored the code and pushed it into the `actionsBuilder` which just exposes a `Build` method. These two methods were renamed to `buildForNewConfig` and `buildForOldConfig`, since it's obvious it's building actions, this makes the naming a bit better, but more importantly, the code is separated from the rest of the `Service` methods.